### ISSUE TITLE
Error in 'BPNet-OSKN'

### ISSUE THREAD

**Issue by IshikaLuthra** [kipoi/kipoi#649]
Hi,

I am having issues importing BPNet with python.

```
model = kipoi.get_model('BPNet-OSKN')

Using downloaded and verified file: /arc/home/iluthra/.kipoi/models/BPNet-OSKN/downloaded/model_files/model_file/bbe883baef261877bfad07d05feb627d
Traceback (most recent call last):
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_utils/utils.py", line 139, in load_obj
    module = imp.load_module(module_name, fp, pathname, description)
  File "/arc/software/spack-0.14.0-110/opt/spack/linux-centos7-broadwell/gcc-5.4.0/python-3.6.8-vups4jci3tulsghu2jsghmsnjif7dqn7/lib/python3.6/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/arc/software/spack-0.14.0-110/opt/spack/linux-centos7-broadwell/gcc-5.4.0/python-3.6.8-vups4jci3tulsghu2jsghmsnjif7dqn7/lib/python3.6/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/arc/home/iluthra/.kipoi/models/BPNet-OSKN/model.py", line 1, in <module>
    from bpnet.seqmodel import SeqModel
ModuleNotFoundError: No module named 'bpnet'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/model.py", line 196, in get_model
    Mod = load_obj(md.defined_as)
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_utils/utils.py", line 144, in load_obj
    raise ImportError("object {} couldn't be imported. Error {}".format(obj_import, str(e)))
ImportError: object model.BPNetOldSeqModel couldn't be imported. Error No module named 'bpnet'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/model.py", line 204, in get_model
    raise ImportError("Unable to import {}".format(md.defined_as))
ImportError: Unable to import model.BPNetOldSeqModel
```

Any help would be appreciated, Thank you!

**Comment by haimasree** [comment]
Hi Ishika - yes thats correct. Sorry if it is not clear from the website. I can fix it tomorrow. Let me know how it goes.

**Comment by IshikaLuthra** [comment]
Hi Haimasree,

If we have installed kipoi in a Conda environment then how should we go about activating another one within it for each model? Or should we not install kipoi in a Conda environment but in our home directory instead?

Thank you,

Ishika

**Comment by haimasree** [comment]
Hi Ishika, in principle yes. So, the following workflow will be okay.
```
conda create -n py38 python=3.8
conda activate py38
pip install kipoi
kipoi env create BPNet-OSKN -> Installs BPNet-OSKN specific dependencies to a **new** conda environment
kipoi env create Basset -> Installs Basset specific dependencies to a **new** conda environment
conda deactivate  (not necessary but for your peace of mind :))
```

```
conda activate kipoi-BPNet-OSKN
<do your bpnet oskn analysis here>
conda deactivate (not necessary but for your peace of mind :))
```
```
conda activate kipoi-Basset
<do your basset analysis here>
conda deactivate (not necessary but for your peace of mind :))
```

Hope this helps?

**Comment by IshikaLuthra** [comment]
Hi,

I attempted doing it that way and I now get other weird errors. 

for `kipoi env create Basset` I get:

`INFO [kipoi.cli.env] Writing environment file: /tmp/kipoi/envfiles/747fdd8d
INFO [kipoi.cli.env] Loading model: Basset description
INFO [kipoi.sources] Update /arc/home/iluthra/.kipoi/models/
Already up to date.
INFO [kipoi.cli.env] Environment name: kipoi-Basset
INFO [kipoi.cli.env] Output env file: /tmp/kipoi/envfiles/747fdd8d/kipoi-Basset.yaml
INFO [kipoi.cli.env] Done writing the environment file!
INFO [kipoi.cli.env] Creating conda env from file: /tmp/kipoi/envfiles/747fdd8d/kipoi-Basset.yaml


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.11.0

Please update conda by running

    $ conda update -n base -c defaults conda






Traceback (most recent call last):
  File "/home/iluthra/.local/bin/kipoi", line 10, in <module>
    sys.exit(main())
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/__main__.py", line 111, in main
    command_fn(args.command, sys.argv[2:])
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/cli/env.py", line 677, in cli_main
    command_fn(args.command, raw_args[1:])
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/cli/env.py", line 447, in cli_create
    env_db_entry.cli_path = get_kipoi_bin(env)
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_conda/utils.py", line 121, in get_kipoi_bin
    raise ValueError("kipoi is not installed in conda environment: {0}".format(env_name))
`
For BPNET

INFO [kipoi.cli.env] Writing environment file: /tmp/kipoi/envfiles/474820eb
INFO [kipoi.cli.env] Loading model: BPNet-OSKN description
INFO [kipoi.sources] Update /arc/home/iluthra/.kipoi/models/
Already up to date.
INFO [kipoi.cli.env] Environment name: kipoi-BPNet-OSKN
INFO [kipoi.cli.env] Output env file: /tmp/kipoi/envfiles/474820eb/kipoi-BPNet-OSKN.yaml
INFO [kipoi.cli.env] Done writing the environment file!
INFO [kipoi.cli.env] Creating conda env from file: /tmp/kipoi/envfiles/474820eb/kipoi-BPNet-OSKN.yaml

CondaValueError: prefix already exists: /home/iluthra/.conda/envs/kipoi-BPNet-OSKN

`Traceback (most recent call last):
  File "/home/iluthra/.local/bin/kipoi", line 10, in <module>
    sys.exit(main())
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/__main__.py", line 111, in main
    command_fn(args.command, sys.argv[2:])
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/cli/env.py", line 677, in cli_main
    command_fn(args.command, raw_args[1:])
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/cli/env.py", line 443, in cli_create
    kipoi_conda.create_env_from_file(env_file, use_stdout=args.verbose)
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_conda/utils.py", line 127, in create_env_from_file
    return _call_conda(cmd_list, use_stdout=use_stdout, dry_run=dry_run)
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_conda/utils.py", line 196, in _call_conda
    return _call_command("conda", extra_args, use_stdout, return_logs_with_stdout, dry_run=dry_run)
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_utils/utils.py", line 78, in _call_command
    raise Exception("could not invoke {0} \nreturn code:{1}\nadditional info:{2}".format(cmd_list, return_code, "".join(error_out)))
Exception: could not invoke ['conda', 'env', 'create', '--file', '/tmp/kipoi/envfiles/474820eb/kipoi-BPNet-OSKN.yaml'] 
return code:1
additional info:`

I'm not sure why this could be happening? Thank you for your help!

**Comment by haimasree** [comment]
For BPNet the error is right in the message you wrote above - ```CondaValueError: prefix already exists: /home/iluthra/.conda/envs/kipoi-BPNet-OSKN```. Remove the directory and try again.

I have no idea about the error you get from Basset. Perhaps show me the content of this  yaml file - INFO [kipoi.cli.env] Creating conda env from file: /tmp/kipoi/envfiles/747fdd8d/kipoi-Basset.yaml? 

In any case add a --verbose flag like so ```kipoi env create Basset --verbose```.  Perhaps some of the dependencies are not getting resolved.

In any case, our CI has not picked up any problem in Basset or BPNet-OSKN. Our test platform is ubuntu 2021.04 and we use conda 4.10.3. I am not sure if the older version of conda is causing a problem.

**Comment by IshikaLuthra** [comment]
Hi,

I tried removing the env and it spends a lot of time but finally outputs this error:

```
kipoi env create BPNet-OSKN
INFO [kipoi.cli.env] Writing environment file: /tmp/kipoi/envfiles/4b995dfc
INFO [kipoi.cli.env] Loading model: BPNet-OSKN description
INFO [kipoi.sources] Update /arc/home/iluthra/.kipoi/models/
Already up to date.
INFO [kipoi.cli.env] Environment name: kipoi-BPNet-OSKN
INFO [kipoi.cli.env] Output env file: /tmp/kipoi/envfiles/4b995dfc/kipoi-BPNet-OSKN.yaml
INFO [kipoi.cli.env] Done writing the environment file!
INFO [kipoi.cli.env] Creating conda env from file: /tmp/kipoi/envfiles/4b995dfc/kipoi-BPNet-OSKN.yaml


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.11.0

Please update conda by running

    $ conda update -n base -c defaults conda




Traceback (most recent call last):
  File "/home/iluthra/.local/bin/kipoi", line 10, in <module>
    sys.exit(main())
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/__main__.py", line 111, in main
    command_fn(args.command, sys.argv[2:])
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/cli/env.py", line 677, in cli_main
    command_fn(args.command, raw_args[1:])
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi/cli/env.py", line 447, in cli_create
    env_db_entry.cli_path = get_kipoi_bin(env)
  File "/home/iluthra/.local/lib/python3.6/site-packages/kipoi_conda/utils.py", line 121, in get_kipoi_bin
    raise ValueError("kipoi is not installed in conda environment: {0}".format(env_name))
ValueError: kipoi is not installed in conda environment: kipoi-BPNet-OSKN

```

Why could this be happening?

Ishika

**Comment by haimasree** [comment]
This error seems identical to what you wrote for Basset above. Like I said try adding a verbose flag or send me the content of the environment file (explained above) or give me the platform information.

If you would like to use containers there is a singularity container available and information on how to use it under the singularity tab.

**Comment by IshikaLuthra** [comment]
Hi Haimasree,

Sorry for the delay. We were able to get it working via command line and when we run BPNet-OSKN in python we are able to do it when we launch python and run it line-by-line but it does not work when submitting a python script with the same code. We are trying to figure out why this may be happening.

I will keep you updated, Thank you!

Ishika

**Comment by haimasree** [comment]
Thats very strange. It is a bit of a weird suggestion but if you rename your script to something different can you replicate the behavior?

**Comment by IshikaLuthra** [comment]
Hi Haimasree,

The tech support person for our cluster helping me found this error:

`ImportError: Unable to import model.BPNetOldSeqModel
>>>

So it turns out if you look at the model BPNet-OSKN that the kipoi installs is not the correct version. 

and inside the following file. 

/arc/home/iluthra/.kipoi/models/BPNet-OSKN/model.py 
it was calling for the wrong BPNetSeqModel. 

I updated this model the the latest version and modified line 17 of the model.py to 

class BPNetOldSeqModel(BaseModel, SeqModel): 
`
We were able to run it with this change once, but the installation keeps breaking. Do you have any thoughts on this?

Kind Regards,

Ishika

**Comment by haimasree** [comment]
Hi Ishika, I am not too sure what you mean by old and new. Kipoi uses [this](https://github.com/kundajelab/bpnet/tree/0cb7277b736260f8b4084c9b0c5bd62b9edb5266) commit of the BPNet-OSKN repository. You can  check that by noticing kipoi installs BPNet-OSKN specifically from this commit [here](https://github.com/kipoi/models/blob/master/BPNet-OSKN/model.yaml#L51). If the model weights or anything else have changed recently it will not be reflected in the version available in Kipoi. This pinning is important to ensure reproducibility. You may try to install BPNET-OSKN from the master branch by changing your local model.yaml (typically in `~/.kipoi/models/BPNet-OSKN/model.yaml`) to `git+https://github.com/kundajelab/bpnet.git@master`  but I am afraid I cant guarantee it will work. However, I still recommend you send me the conda environment file and the os version so that I can try it out and see if I can reproduce your error. Otherwise, I dont have much to go on.

**Comment by haimasree** [comment]
Another detail: If you want to retain changes you have made to `~/.kipoi/models/..` you need to set `auto_update` to `False` in `~/.kipoi/config.yaml`.

```
model_sources:
    kipoi:
        type: git
        remote_url: https://github.com/kipoi/models.git
        local_path: /Users/b260/.kipoi/models
        **auto_update: False**
        use_lfs: false
    github-permalink:
        type: github-permalink
        local_path: /Users/b260/.kipoi/github-permalink/
```

**Comment by IshikaLuthra** [comment]
Hi Haimasree,

Thank you for all your help! Our technical support person found that the error was being caused by the python script name I set (bpnet.py). This was causing confusion under the hood when we were creating the conda environment since that name must be used somewhere else. Once we changed the name it cleared up the issues. 

Everything seems to be work now! I really appreciate your help during the debugging :)

Kind Regards,

Ishika
